### PR TITLE
Logan Medina's Finished Book Project

### PR DESCRIPTION
**Pull request by medinal**

### CODE REVIEW

**Comment by medinal** [comment]
5/5 5/5 5/5 for everything. It went well.

File: controllers/BooksIndexController.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  .controller('BooksIndexController', BooksIndexController);
+
+var db = "https://super-crud.herokuapp.com/books"
+
+BooksIndexController.$inject = ['$http'];
+function BooksIndexController(   $http ){
+
+  var vm = this;
+
+  vm.books = [];
```
Comment (wilkdasilk) [inline controllers/BooksIndexController.js:12]: local database?

File: controllers/BooksShowController.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+function BooksShowController(   $http,   $routeParams,  $location){
+
+  var vm = this;
+
+  vm.book = {};
+
+  vm.editedBook = {};
+
+  vm.editing = function(){
+    vm.editedBook = Object.assign({}, vm.book);
```
Comment (wilkdasilk) [inline controllers/BooksShowController.js:17]: cool object method

File: controllers/BooksShowController.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  vm.editBook = function(){
+    $http({
+      method: 'PUT',
+      url: `${db}/${$routeParams.id}`,
```
Comment (wilkdasilk) [inline controllers/BooksShowController.js:23]: no need for quotes or template literals, just ... 
db + $routeParams.id,

if needed, add a slash to the end of your db url declaration

File: controllers/BooksShowController.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      vm.book = response.data
+    }), function errorCallback(response){
+      console.log(response);
+    }
+  }
+
+  vm.deleteBook = function(){
+    $http({
+      method: 'DELETE',
+      url: `${db}/${$routeParams.id}`,
```
Comment (wilkdasilk) [inline controllers/BooksShowController.js:35]: see above comment

File: index.html [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  </head>
+  <body>
+    <div class="jumbotron">
+      <div class="container">
+        <a href="/"><h1 class="text-right">BookMe!</h1></a>
+      </div>
+    </div>
+
+    <div ng-view>
```
Comment (wilkdasilk) [inline index.html:36]: user-experience - show something initially to let the user know content is loading

File: public/scripts/app.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    .when('/', {
+      templateUrl: '/views/books.html',
+      controllerAs: 'booksIndexCtrl',
+      controller: 'BooksIndexController'
+    })
+    .when('/books/:id', {
+      templateUrl: '/views/books-show.html',
+      controllerAs: 'booksShowCtrl',
+      controller: 'BooksShowController'
+    })
```
Comment (wilkdasilk) [inline public/scripts/app.js:17]: semicolon?

File: public/styles/styles.css [hunk 7]
```diff
@@ -0,0 +1,47 @@
+h1 {
+  font-family: 'Lora', serif;
+  animation: type 4s steps(60, end);
```
Comment (wilkdasilk) [inline public/styles/styles.css:3]: what is this supposed to do?

File: public/styles/styles.css [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+a:visited, a:focus {
+  text-decoration: none;
+  color: #704418;
+}
+
+img {
+  height: 300px;
+}
+
+.individual {
```
Comment (wilkdasilk) [inline public/styles/styles.css:45]: class name could be more descriptive

File: public/styles/styles.css [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+a {
+  color: #704418;
+}
+
+a:hover {
+  text-decoration: none;
+  color: #A5703A;
+}
+
+a:visited, a:focus {
```
Comment (wilkdasilk) [inline public/styles/styles.css:36]: css-reset?

File: views/books-show.html [hunk 10]
```diff
@@ -0,0 +1,27 @@
+<div class="container">
+  <hr>
+    <div class="clearfix">
```
Comment (wilkdasilk) [inline views/books-show.html:3]: might be able to stop shuffle behavior by starting a new bootstrap row instead of clearfix

File: views/books-show.html [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+      <div class="col-md-6">
+        <h3><b>Title:</b> <span ng-hide="editing"> {{ booksShowCtrl.book.title }}</span></h3>
```
Comment (wilkdasilk) [inline views/books-show.html:5]: best practice: separate style and semantic meaning - aka no `<b>`, use a span with a class instead

File: views/books.html [hunk 12]
```diff
@@ -0,0 +1,29 @@
+<div class="container">
+  <input class="pull-right" type="search" ng-model="q" placeholder="filter books..." aria-label="filter books" />
+  <form ng-show="add" ng-submit="booksIndexCtrl.createBook(); add = false">
```
Comment (wilkdasilk) [inline views/books.html:3]: easuer to understand if you use a different show-hide variable name - ie adding instead of add

"show while (adding)" makes more sense as a state than "show while (add)"

File: views/books.html [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    <input type="text" class='form-control' ng-model="booksIndexCtrl.newBook.title" placeholder='New book title' >
+    <input type="text" class='form-control' ng-model="booksIndexCtrl.newBook.author" placeholder='New book author' >
+    <input type="text" class='form-control' ng-model="booksIndexCtrl.newBook.releaseDate" placeholder='New book release date' >
+    <input type="text" class='form-control' ng-model="booksIndexCtrl.newBook.image" placeholder='Image Url' >
+    <input class="btn btn-default" type="submit" id="submit" value="Save">
+    <button type="button" ng-show="add" class="btn btn-danger" ng-click="add=false">Close</button>
+  </form>
+  <input ng-hide="add" ng-click="add = true" class="btn btn-default" value="Add book">
+  <hr>
+  <div ng-repeat="book in booksIndexCtrl.books | orderBy : 'author' | filter:q as results">
```
Comment (wilkdasilk) [inline views/books.html:13]: not sure you need to give an alias to q. do you use the alias anywhere?

**Review by wilkdasilk — COMMENTED** [review]
App functions well, good stretch challenges, just a few ways to simplify your code or make it easier to work with going forward

File: controllers/BooksIndexController.js (hunk 1 above)
Comment (mnfmnfm) [reply controllers/BooksIndexController.js:12]: This is fine! You probably need some local storage of the things you've fetched from the db.

File: controllers/BooksIndexController.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
+angular
+  .module('booksApp')
+  .controller('BooksIndexController', BooksIndexController);
+
+var db = "https://super-crud.herokuapp.com/books"
```
Comment (mnfmnfm) [inline controllers/BooksIndexController.js:5]: missing semicolon

File: controllers/BooksIndexController.js [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+  vm.newBook = {};
+
+  vm.createBook = function(){
+    $http({
+      method: 'POST',
+      url: db,
+      data: vm.newBook
+    }).then(function successCallback(response){
+      vm.books.push(response.data)
```
Comment (mnfmnfm) [inline controllers/BooksIndexController.js:22]: missing so many semicolons

File: controllers/BooksIndexController.js [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    }), function errorCallback(response){
+      console.log(response)
+    }
+  }
+
+  $http({
+    method: 'GET',
+    url: db
+  }).then(function successCallback(response) {
+    vm.books = response.data.books
```
Comment (mnfmnfm) [inline controllers/BooksIndexController.js:33]: why are there no semicolons 😭

File: controllers/BooksShowController.js [hunk 17]
```diff
@@ -0,0 +1,52 @@
+angular
+  .module('booksApp')
+  .controller('BooksShowController', BooksShowController);
+
+var db = "https://super-crud.herokuapp.com/books"
```
Comment (mnfmnfm) [inline controllers/BooksShowController.js:5]: 😭

File: controllers/BooksShowController.js (hunk 3 above)
Comment (mnfmnfm) [reply controllers/BooksShowController.js:23]: **reviewer disagreement whooo**
I agree that you could just add the slash to your url declaration and use +, but I also like your use of template literals. /shrug

File: controllers/BooksShowController.js (hunk 17 above)
Comment (mnfmnfm) [reply controllers/BooksShowController.js:5]: Also, this looks like a constant to me, so I would suggest using ALL_CAPS_AND_UNDERSCORES to name it. (i.e. DB or DB_URL rather than db)

File: controllers/BooksShowController.js [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+    vm.editedBook = Object.assign({}, vm.book);
+  }
+
+  vm.editBook = function(){
+    $http({
+      method: 'PUT',
+      url: `${db}/${$routeParams.id}`,
+      data: vm.editedBook
+    }).then(function successCallback(response){
+      vm.book = response.data
```
Comment (mnfmnfm) [inline controllers/BooksShowController.js:26]: 😭

File: controllers/BooksShowController.js [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+      console.log(response);
+    }
+  }
+
+  vm.deleteBook = function(){
+    $http({
+      method: 'DELETE',
+      url: `${db}/${$routeParams.id}`,
+    }).then(function successCallback(response){
+      $location.path('/')
```
Comment (mnfmnfm) [inline controllers/BooksShowController.js:37]: 😭

File: controllers/BooksShowController.js [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+    }), function errorCallback(response){
+      console.log(response)
```
Comment (mnfmnfm) [inline controllers/BooksShowController.js:39]: 😭

File: controllers/BooksShowController.js [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+    }
+  }
+
+  $http({
+    method: 'GET',
+    url: `${db}/${$routeParams.id}`,
+  }).then(function successCallback(response) {
+    vm.book = response.data
```
Comment (mnfmnfm) [inline controllers/BooksShowController.js:47]: 😭

File: controllers/BooksShowController.js [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+  }), function errorCallback(response) {
```
Comment (mnfmnfm) [inline controllers/BooksShowController.js:48]: This straight-up isn't working; you close your .then() function call with this close paren, and I'm amazed you don't get a syntax error.

File: index.html [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+    <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.6.2/angular-route.js"></script>
+
+    <!-- APPLICATION SCRIPTS -->
+    <script src="public/scripts/app.js"></script>
+    <script src="controllers/BooksIndexController.js"></script>
+    <script src="controllers/BooksShowController.js"></script>
+
+  </head>
+  <body>
+    <div class="jumbotron">
```
Comment (mnfmnfm) [inline index.html:30]: Why do you have both this jumbotron and container? Do they do nice things for you, or did you just copy/paste them from other assignments?

File: index.html [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+
+    <!-- fonts -->
+    <link href="https://fonts.googleapis.com/css?family=Lora|Open+Sans" rel="stylesheet">
+
+    <!-- VENDOR SCRIPTS -->
+    <script src="https://code.angularjs.org/1.6.2/angular.min.js"></script>
+    <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.6.2/angular-route.js"></script>
+
+    <!-- APPLICATION SCRIPTS -->
+    <script src="public/scripts/app.js"></script>
```
Comment (mnfmnfm) [inline index.html:24]: Because you use relative links instead of more-absolute links, this breaks if you refresh any book-detail page. Make sure to start your urls with /.

File: public/scripts/app.js (hunk 6 above)
Comment (mnfmnfm) [reply public/scripts/app.js:17]: 😭

File: views/books.html (hunk 12 above)
Comment (mnfmnfm) [reply views/books.html:3]: more controversy: i don't care

File: views/books.html [hunk 25]
```diff
@@ -0,0 +1,29 @@
+<div class="container">
+  <input class="pull-right" type="search" ng-model="q" placeholder="filter books..." aria-label="filter books" />
+  <form ng-show="add" ng-submit="booksIndexCtrl.createBook(); add = false">
+    <input type="text" class='form-control' ng-model="booksIndexCtrl.newBook.title" placeholder='New book title' >
```
Comment (mnfmnfm) [inline views/books.html:4]: Should use labels rather than placeholders to inform people about what to put in boxes; recall that placeholders aren't guaranteed to be shown on all devices.

**Review by mnfmnfm — COMMENTED** [review]
You got lucky on your URLs, and lots of missing semicolons 😭 but generally looks good.
